lib/puppet/data_providers.rb

```diff
@@ -17,12 +17,15 @@ def self.lookup_in_environment(name, scope, merge)
   MODULE_NAME = 'module_name'.freeze
 
   def self.lookup_in_module(name, scope, merge)
-    # Do not attempt to do a lookup in a module if evaluated code is not in a module
-    # which is detected by checking if "MODULE_NAME" exists in scope
-    return nil unless scope.exist?(MODULE_NAME)
+    # Do not attempt to do a lookup in a module unless the name is qualified.
+    qual_index = name.index('::')
+    throw :no_such_key if qual_index.nil?
+    module_name = name[0..qual_index-1]
 
     assert_loaded()
     adapter = Puppet::DataProviders::DataAdapter.adapt(Puppet.lookup(:current_environment))
-    adapter.module_provider(scope[MODULE_NAME]).lookup(name, scope, merge)
+    data_provider = adapter.module_provider(module_name, true)
+    throw :no_such_key if data_provider.nil?
+    data_provider.lookup(name, scope, merge)
   end
```

Comment:
Not sure what you mean. Throwing a :no_such_key exception doesn't imply failure. It just meanst that no value was found for the given key. The alternative is to say that the key was found and its value is `nil` which is incorrect IMO.